src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,182 @@
+
+            // Get the set of results the request said were previously reported.
+            var previousResults = GetPreviousResults(diagnosticsParams);
+
+            var documentToPreviousResultId = new Dictionary<Document, string?>();
+            if (previousResults != null)
+            {
+                // Go through the previousResults and check if we need to remove diagnostic information for any documents
+                foreach (var previousResult in previousResults)
+                {
+                    if (previousResult.TextDocument != null)
+                    {
+                        var document = _solutionProvider.GetDocument(previousResult.TextDocument);
+                        if (document == null)
+                        {
+                            // We can no longer get this document, return null for both diagnostics and resultId
+                            progress.Report(CreateReport(previousResult.TextDocument, diagnostics: null, resultId: null));
+                        }
+                        else
+                        {
+                            // Cache the document to previousResultId mapping so we can easily retrieve the resultId later.
+                            documentToPreviousResultId[document] = previousResult.PreviousResultId;
+                        }
+                    }
+                }
+            }
+
+            // Go through the documents that we need to process and call XamlPullDiagnosticService to get the diagnostic report
+            foreach (var document in GetDocuments(context))
+            {
+                SourceText text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                var documentId = ProtocolConversions.DocumentToTextDocumentIdentifier(document);
+
+                // If we can get a previousId of the document, use it, 
+                // otherwise use null as the previousId to pass into the XamlPullDiagnosticService
+                var previousResultId = documentToPreviousResultId.ContainsKey(document) ? documentToPreviousResultId[document] : null;
+
+                // Call XamlPullDiagnosticService to get the diagnostic report for this document.
+                // We will compute what to report inside XamlPullDiagnosticService, for example, whether we should keep using the previousId or use a new resultId,
+                // and the handler here just return the result get from XamlPullDiagnosticService.
+                var diagnosticReport = await _xamlDiagnosticService.GetDiagnosticReportAsync(document, previousResultId, cancellationToken).ConfigureAwait(false);
+                progress.Report(CreateReport(
+                            documentId,
+                            ConvertToVSDiagnostics(diagnosticReport.Diagnostics, document, text),
+                            diagnosticReport.ResultId));
+            }
+
+            return null;
```

Comment:
shouldn't this return `progress.GetValues();` here?

Reply:
I see you return GetValues() in your handler but from the wiki it tells me to return null - 

[https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki/11004/Error-List-Remote-Pull-Model-Diagnostic?anchor=language-server-side](https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki/11004/Error-List-Remote-Pull-Model-Diagnostic?anchor=language-server-side)
[JsonRpcMethod(MSLSPMethods.DocumentPullDiagnosticName, UseSingleObjectParameterDeserialization = true)]
public object GetDocumentDiagnostics(DocumentDiagnosticsParams parameters, CancellationToken token)
{
    // [...]
    parameters.PartialResultToken.Report(diagnosticReports);
    // [...]
    return null; // if diagnostics are sent through PartialResultToken, returning a value isn't necessary.
}

I think my concern is if we have reported diagnostics already, will return progress.GetValues() report dup information?

---
In reply to: [516285646](https://github.com/dotnet/roslyn/pull/49145#discussion_r516285646) [](ancestors = 516285646)

Reply:
> I think my concern is if we have reported diagnostics already, will return progress.GetValues() report dup information?

No.  It's either/or.  If the caller is passing in an IProgress, then only that is used.  If they don't, then the results are passed as the array at the end.

Reply:
Changed it to return GetValues()